### PR TITLE
Add two biome experiments and X33 data for paper

### PR DESCRIPTION
**Pull request by EricXiong01**
This pull request makes several important updates to cluster configuration files and experiment sweep settings, focusing on resource allocation adjustments and experiment hyperparameter simplification. It also adds new cluster configuration files for various GPU job types and introduces a new script for plotting learning curves. The most significant changes are grouped below:

**Cluster resource allocation updates:**

* Reduced the number of CPU cores from 16 to 8 and doubled `mem_per_core` from 8G to 16G in `vulcan-cpu-1h.json`, `vulcan-cpu-2h.json`, and `vulcan-cpu-6h.json` to better match workload requirements. [[1]](diffhunk://#diff-c0911faa3f3288837110d72cae20739a98a2ad058de3ed7c94dc3cfbd7cb4041L6-R7) [[2]](diffhunk://#diff-afedb6c606d4c70c1342ac473bfef706556b04e6a64b5261fd5682d8727fac38L6-R7) [[3]](diffhunk://#diff-b90c58a5d453b05f5b2369a8118bc03f17980e14f6133e2590d71719e88ba3dcL6-R7)
* Increased `tasks_per_core` and `tasks_per_vmap` from 150/128 to 300 in `vulcan-gpu-vmap-1h.json` and `vulcan-gpu-vmap-2h.json` for higher parallelism. [[1]](diffhunk://#diff-cb6b6e18976140540dcca6012e8a18b9ff765d347e85b7a2e643d569b96eb205L10-R11) [[2]](diffhunk://#diff-7dd5d8d42fc8eca2c5cb6e50a63a61bc98c9930e27053a2aa51a971970666646L10-R11)

**New cluster configuration files:**

* Added several new GPU cluster configuration JSONs for various durations and task sizes, including: `vulcan-gpu-vmap-1h-35.json`, `vulcan-gpu-vmap-1h-70.json`, `vulcan-gpu-vmap-20m-8.json`, `vulcan-gpu-vmap-3h-5.json`, `vulcan-gpu-vmap-3h-15.json`, `vulcan-gpu-vmap-45m-high.json`, `vulcan-gpu-vmap-6h-15.json`, `vulcan-gpu-vmap-6h-30.json`, and `vulcan-gpu-vmap-30m-high-512.json`. These provide more granular resource options for different experiment needs. [[1]](diffhunk://#diff-6e8b29b25140a95afe9b227a9103c9e8a833c367799f01d492b262c54adc0917R1-R12) [[2]](diffhunk://#diff-3566244eb7841de0cfda574b94e9578f1573f430b0d37b1ccb298747558685b8R1-R12) [[3]](diffhunk://#diff-eb84458837d2785594022a6db85dc6ee79edcfb4a05e4bd3749dbc4e41294d6bR1-R12) [[4]](diffhunk://#diff-521b9e75816619ca40688e7bae7711446aff7fd73d3bb18a4fbd8e0db7e6c755R1-R12) [[5]](diffhunk://#diff-a1563a48033118fe88affe6c75fe9e228fa4e1723f1270eed72718a2f6639e39R1-R12) [[6]](diffhunk://#diff-2c05b77957cb3d575ea9993b235084dbc5f773a3f3e1dc12d61934b44503f3faR1-R12) [[7]](diffhunk://#diff-da61f411d43bcabe0a96bd50ce7d16a8b382d3c8f8306e42e187cf1bff92a857R1-R12) [[8]](diffhunk://#diff-5748f2e138cd4bf1a5f5e57020b34df229133945a0f31e6ad12c848d4f9dcdebR1-R12) [[9]](diffhunk://#diff-0554ae5aba2f53d62260b44a7536aa32dac479da27e4ce05d33a2e42bb1e768eR1-R12)

**Experiment sweep and hyperparameter simplification:**

* In multiple `ActorCriticMLP` JSONs, changed the optimizer `beta2` parameter from a list (`[0.999, 0.9]`) to a single value (`0.999`), simplifying the sweep and reducing the hyperparameter search space. [[1]](diffhunk://#diff-14111572ed476ea72a069b7c8d82200393bc7a87df173ccc373836a3b4d624b1L41-R41) [[2]](diffhunk://#diff-14111572ed476ea72a069b7c8d82200393bc7a87df173ccc373836a3b4d624b1L52-R52) [[3]](diffhunk://#diff-4fed1c4605b7acbf347dcbfc11df7c6a7cdc1a93418d67bbde5c1fe6ddeb51b5L41-R41) [[4]](diffhunk://#diff-4fed1c4605b7acbf347dcbfc11df7c6a7cdc1a93418d67bbde5c1fe6ddeb51b5L52-R52) [[5]](diffhunk://#diff-062598eafe0e46e1f3912652e1802eacc0f4f469dd817e8cb0c91ab2b7de1ad0L38-R38) [[6]](diffhunk://#diff-062598eafe0e46e1f3912652e1802eacc0f4f469dd817e8cb0c91ab2b7de1ad0L49-R49) [[7]](diffhunk://#diff-031de9175db2b8bca9adbaadca361eafb1de4e05fe4d3ed671de021405908da1L36-R36) [[8]](diffhunk://#diff-031de9175db2b8bca9adbaadca361eafb1de4e05fe4d3ed671de021405908da1L47-R47)
* In `DQN.json`, revised hyperparameters: made `epsilon` a sweep over `[0.05, 0.1, 0.25]`, fixed `target_refresh` to 128, simplified optimizer settings, and switched the representation to `ForagerNet` with layer norm. Also changed the observation type to `"color"`. [[1]](diffhunk://#diff-2d9a16390d047a40ca5b86d82205e4aaf0f215c753fd982e38d6f1bc5e5f1d1fL7-R15) [[2]](diffhunk://#diff-2d9a16390d047a40ca5b86d82205e4aaf0f215c753fd982e38d6f1bc5e5f1d1fL30-R37)

**Experiment analysis and plotting:**

* Added a new script `learning_curve2.py` for generating and saving learning curve plots with bootstrap confidence intervals, supporting improved experiment analysis and visualization.

### CODE REVIEW

**Comment by steventango** [comment]
Claude Code Review:

# Code Review — PR #159: Add two biome experiments and X33 data for paper

## Overview

A large PR (~11.6k +, 850 −, 294 files) bundling four streams of work:
1. **Cluster configs** — adjusted CPU memory layout (16×8G → 8×16G) and added 9 new GPU partition configs (`vulcan-gpu-vmap-*`) for finer-grained slurm scheduling.
2. **Experiment configs** — pruned hyperparameter sweeps (e.g. `beta2: [0.999,0.9]` → `0.999`), added X33 experiments for the paper (DRQN, PT_DQN, ActorCriticMLP variants), and two new collections (`X33`, `XN33`).
3. **Plotting** — significant rework of `src/learning_curve.py` (bar plots, color/legend/FOV controls, gradient palettes), plus new `learning_curve2.py` scripts.
4. **Core code** — generalized DRQN/recurrent infra to support RTU carries (tuple), added optional update-skipping path for accumulated `update_freq>1`, refactored `ForagerGRUNetReLU` to use `pre_core_layers`/`post_core_layers` (with backward-compat aliases for `pre_gru_layers`), added new `ForagerRTUNetReLU` and `RTU` modules, added Reset/S&P regularization to `rtu_ppo.py`.

## Issues / Risks

### Bugs (must fix)

- **`src/learning_curve.py:461` — undefined `lookup_color`.** The function definition was removed (replaced by `args.color_algs`/`args.filter_alg_apertures` dict-comp branches), but the `elif args.filter_algs:` branch still calls `lookup_color(alg, fallback_idx)`. Any invocation with `--filter-algs` (without `--color-algs` or `--filter-alg-apertures`) will hit `NameError`.

- **`src/rtu_ppo.py:5-10` — duplicate imports.** The patch added `import socket`, `import time`, and `import utils.jax_compat` near the top, but didn't remove the existing copies further down. Result on the branch: `import socket` and `import time` each appear twice, and `import utils.jax_compat` appears at lines 6 and 31. Harmless at runtime but a lint failure / code smell.

### Concerns

- **`src/continuing_main.py:240` — `save_every` default changed `100_000 → 10_001_000`.** That's a 100× delay. If `experiment.save_every` isn't explicitly set in a config, runs effectively no longer checkpoint until the very end. The original `100_000` looks safer as a default; if the intent was "checkpoint once at end for short runs," it should be a flag, not a default.

- **`src/continuing_main.py:33` — `UNROLL = 4 → 1`.** Reduces compile time/memory but may regress step throughput. Worth a perf note in the PR body.

- **`src/rtu_ppo.py:1289-1297` — commented-out `freeze_steps_end` block left as dead code.** Either delete it or restore it; the inline commented block will rot.

- **`src/representations/networks.py:430-450` (`RTU.__call__`) — `is_target` skips the reset-to-initial-state branch.** `if not is_target:` means target-network rollouts ignore episode boundaries. That's plausible for burn-in but is asymmetric vs. the GRU path (which always resets). Confirm this is intentional and document — silent divergence between online/target carries during DRQN burn-in is a class of subtle bug.

- ~~**`src/representations/networks.py` — `ForagerRTUNetReLU` constructor accepts `use_layernorm` and `learn_initial_h` but never uses them on the recurrent core itself.**~~ **Correction:** `ForagerGRUNetReLU` does the same thing (only applies LN to conv/embedding and pre/post MLPs, not the GRU core), so the RTU path is consistent. `learn_initial_h` is genuinely unused in the RTU path (RTU always zero-inits) but that's a minor asymmetry, not a silent bug.

- **`src/representations/networks.py:1502, 1523` — `RTU` is instantiated inside `__call__` rather than `__init__`** (`rtu = RTU(hidden=core_in.shape[-1], …, name="rtu")`). Module construction inside the call method is allowed by Haiku but unusual; it ties hidden width to the runtime tensor shape. Consider moving to `__init__` once `core_in` width is determinable from config (it is — `hidden + scalars` or similar).

- **`src/algorithms/nn/DRQN.py` — `_extract_carry`/`_carry_for_buffer`/`_split_carry`/`_set_carry_start` helpers each branch on `isinstance(carry, tuple)`.** This works but threads "is it RTU or GRU" through every recurrent operation. Worth considering a small `Carry` adapter (or just defining `extract/buffer/split/set_start` once, parameterized) to keep DRQN, NNAgent, continuing_main aligned.

- **`scripts/slurm.py:5-9` — old line `sys.path.append(os.getcwd() + "/src")` left commented out.** Delete; the new `Path(__file__).resolve().parents[1] / "src"` block is correct.

- **`src/learning_curve.py:341-450` — figure-sizing branches are heavily duplicated** between the `num_metrics==1` and `else` arms, with the same `if args.horizontal_bars and args.legend_on_bar: …` block repeated four times. Strong refactor candidate (extract `compute_figsize(num_metrics, hue_order, args)`).

- **`src/learning_curve.py:558-570` — frozen-base regex.** `re.sub(r"[-_ ]?frozen(?:[-_ ][a-zA-Z0-9]+)?", ...)` will eat `frozen` substrings unconditionally. If any future label has `frozen` in a non-suffix position, base resolution will silently misfire. A stricter `re.sub(r"_frozen_(1M|5M)$", ...)` is closer to the actual schema in `plotting_utils.py`.

### Minor

- `src/utils/results.py:95` — `dtype.shape[0]` → `dtype.width`. This is the right polars API for `pl.Array`, but assumes a recent polars version (>=0.20). Pin in `pyproject.toml` if not already.
- `src/representations/networks.py:1573` — fixes a real bug (`a = jnp.zeros((N, T, self.number_of_scalars))` → `self.number_of_actions`).
- `src/plotting_utils.py` — `_frozen_5M` now collapses to `(F)` while `_frozen_1M` is `(1M)`. Asymmetric; intentional for paper figures? Worth a comment.
- The 9 new cluster JSONs are near-identical; a generator template would scale better than hand-maintaining each.

## Test Coverage

No tests touched. Given the carry-shape changes in DRQN and the new RTU recurrent path, an end-to-end smoke test (one short run per algo: `DQN`, `DRQN`, `RTU-PPO`, `PT_DQN`) before merging would catch shape/serialization regressions that JIT errors won't surface until real runs start.

## Security

Nothing notable — no auth/IO changes, no new external network calls. Cluster configs are project-internal.

## Recommendation

**Request changes.** The `lookup_color` NameError is a hard break for an existing CLI flag, and the duplicate imports / dead `freeze_steps_end` block / 100× `save_every` default change should each be addressed before merging. The recurrent-core refactor is a good direction but the `is_target`-skips-reset behavior in the RTU path needs clarification.

File: experiments/X32-ForagaxSquareWaveTwoBiome-v11/foragax/ForagaxSquareWaveTwoBiome-v11/make_frozen_5m_configs.py [hunk 1]
```diff
@@ -34,7 +31,7 @@ def main():
 
         data["agent"] = data["agent"] + SUFFIX
         data["metaParameters"]["freeze_steps"] = 5000000
-        data["metaParameters"]["freeze_steps_end"] = 6000000
+        # data["metaParameters"]["freeze_steps_end"] = 6000000
```
Comment (steventango) [inline experiments/X32-ForagaxSquareWaveTwoBiome-v11/foragax/ForagaxSquareWaveTwoBiome-v11/make_frozen_5m_configs.py]:
```suggestion
```

File: experiments/X33-ForagaxSquareWaveTwoBiome-v11/foragax/ForagaxSquareWaveTwoBiome-v11/make_frozen_5m_configs.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+        dst = src.with_name(src.stem + SUFFIX + src.suffix)
+
+        with src.open("r", encoding="utf-8") as f:
+            data = json.load(f)
+
+        data["agent"] = data["agent"] + SUFFIX
+        data["metaParameters"]["freeze_steps"] = 5000000
+        # data["metaParameters"]["freeze_steps_end"] = 6000000
+
```
Comment (steventango) [inline experiments/X33-ForagaxSquareWaveTwoBiome-v11/foragax/ForagaxSquareWaveTwoBiome-v11/make_frozen_5m_configs.py]:
```suggestion

```

File: src/rtu_ppo.py [hunk 3]
```diff
@@ -1148,12 +1298,12 @@ def main():
...
-            freeze_key = jax.random.PRNGKey(seed + 42)
-            freeze_steps_end = hypers["freeze_steps_end"]
-            freeze_steps = jax.random.randint(freeze_key, (), freeze_steps, freeze_steps_end + 1)
-            print(f"Freeze steps sampled to {freeze_steps}")
+        # if "freeze_steps_end" in hypers:
+        #     # TODO: this is to match the key for the env reset, that is the jax.random.PRNGKey(seed)[1] is used to reset the env
+        #     freeze_key = jax.random.PRNGKey(seed + 42)
+        #     freeze_steps_end = hypers["freeze_steps_end"]
+        #     freeze_steps = jax.random.randint(freeze_key, (), freeze_steps, freeze_steps_end + 1)
+        #     print(f"Freeze steps sampled to {freeze_steps}")
```
Comment (steventango) [inline src/rtu_ppo.py]:
```suggestion

```
